# Assignment 01. Exploratory data analysis with Pandas
## Practice. Analyzing "Titanic" passengers

**Fill in the missing code ("You code here")

In [1]:
import numpy as np
import pandas as pd
from matplotlib import pyplot as plt
# Graphics in SVG format are more sharp and legible
%config InlineBackend.figure_format = 'svg'
pd.set_option("display.precision", 2)

**Read data into a Pandas DataFrame**

In [2]:
data = pd.read_csv('../../data/titanic_train.csv',
                  index_col='PassengerId')

**First 5 rows**

In [3]:
data.head(5)

Survived  Pclass  \
PassengerId                     
1                   0       3   
2                   1       1   
3                   1       3   
4                   1       1   
5                   0       3   

                                                          Name     Sex   Age  \
PassengerId                                                                    
1                                      Braund, Mr. Owen Harris    male  22.0   
2            Cumings, Mrs. John Bradley (Florence Briggs Th...  female  38.0   
3                                       Heikkinen, Miss. Laina  female  26.0   
4                 Futrelle, Mrs. Jacques Heath (Lily May Peel)  female  35.0   
5                                     Allen, Mr. William Henry    male  35.0   

             SibSp  Parch            Ticket   Fare Cabin Embarked  
PassengerId                                                        
1                1      0         A/5 21171   7.25   NaN        S  
2                1      0          PC 17599  71.28   C85        C  
3                0      0  STON/O2. 3101282   7.92   NaN        S  
4                1      0            113803  53.10  C123        S  
5                0      0            373450   8.05   NaN        S

In [4]:
data.describe()

Survived  Pclass     Age   SibSp   Parch    Fare
count    891.00  891.00  714.00  891.00  891.00  891.00
mean       0.38    2.31   29.70    0.52    0.38   32.20
std        0.49    0.84   14.53    1.10    0.81   49.69
min        0.00    1.00    0.42    0.00    0.00    0.00
25%        0.00    2.00   20.12    0.00    0.00    7.91
50%        0.00    3.00   28.00    0.00    0.00   14.45
75%        1.00    3.00   38.00    1.00    0.00   31.00
max        1.00    3.00   80.00    8.00    6.00  512.33

**Let's select those passengers who embarked in Cherbourg (Embarked=C) and paid > 200 pounds for their ticker (fare > 200).**

Make sure you understand how actually this construction works.

In [5]:
data[(data['Embarked'] == 'C') & (data.Fare > 200)].head()

Survived  Pclass  \
PassengerId                     
119                 0       1   
259                 1       1   
300                 1       1   
312                 1       1   
378                 0       1   

                                                        Name     Sex   Age  \
PassengerId                                                                  
119                                 Baxter, Mr. Quigg Edmond    male  24.0   
259                                         Ward, Miss. Anna  female  35.0   
300          Baxter, Mrs. James (Helene DeLaudeniere Chaput)  female  50.0   
312                               Ryerson, Miss. Emily Borie  female  18.0   
378                                Widener, Mr. Harry Elkins    male  27.0   

             SibSp  Parch    Ticket    Fare            Cabin Embarked  
PassengerId                                                            
119              0      1  PC 17558  247.52          B58 B60        C  
259              0      0  PC 17755  512.33              NaN        C  
300              0      1  PC 17558  247.52          B58 B60        C  
312              2      2  PC 17608  262.38  B57 B59 B63 B66        C  
378              0      2    113503  211.50              C82        C

**We can sort these people by Fare in descending order.**

In [6]:
data[(data['Embarked'] == 'C') & 
     (data['Fare'] > 200)].sort_values(by='Fare',
                               ascending=False).head()

Survived  Pclass                                   Name     Sex  \
PassengerId                                                                    
259                 1       1                       Ward, Miss. Anna  female   
680                 1       1     Cardeza, Mr. Thomas Drake Martinez    male   
738                 1       1                 Lesurer, Mr. Gustave J    male   
312                 1       1             Ryerson, Miss. Emily Borie  female   
743                 1       1  Ryerson, Miss. Susan Parker "Suzette"  female   

              Age  SibSp  Parch    Ticket    Fare            Cabin Embarked  
PassengerId                                                                  
259          35.0      0      0  PC 17755  512.33              NaN        C  
680          36.0      0      1  PC 17755  512.33      B51 B53 B55        C  
738          35.0      0      0  PC 17755  512.33             B101        C  
312          18.0      2      2  PC 17608  262.38  B57 B59 B63 B66        C  
743          21.0      2      2  PC 17608  262.38  B57 B59 B63 B66        C

**Let's create a new feature.**

In [7]:
def age_category(age):
    '''
    < 30 -> 1
    >= 30, <55 -> 2
    >= 55 -> 3
    '''
    if age < 30:
        return 1
    elif age < 55:
        return 2
    elif age >= 55:
        return 3

In [8]:
age_categories = [age_category(age) for age in data.Age]
data['Age_category'] = age_categories

**Another way is to do it with `apply`.**

In [9]:
data['Age_category'] = data['Age'].apply(age_category)

**1. How many men/women were there onboard?**
- 412 men and 479 women
- 314 men and 577 women
- 479 men and 412 women
- 577 men and 314 women

In [10]:
# You code here
data["Sex"].value_counts()

male      577
female    314
Name: Sex, dtype: int64

* 577 male and 314 female

**2. Print the distribution of the `Pclass` feature. Then the same, but for men and women separately. How many men from second class were there onboard?**
- 104
- 108
- 112
- 125

In [11]:
# Distribution of Pclass
data["Pclass"].describe()

count    891.00
mean       2.31
std        0.84
min        1.00
25%        2.00
50%        3.00
75%        3.00
max        3.00
Name: Pclass, dtype: float64

In [12]:
# Distribution of Pclass depending on Sex
data.loc[:,["Pclass", "Sex"]].groupby("Sex").describe()

Pclass                                     
        count  mean   std  min  25%  50%  75%  max
Sex                                               
female  314.0  2.16  0.86  1.0  1.0  2.0  3.0  3.0
male    577.0  2.39  0.81  1.0  2.0  3.0  3.0  3.0

In [13]:
# Finding the number of men in second class
pd.crosstab(data["Sex"], data["Pclass"])

Pclass    1    2    3
Sex                  
female   94   76  144
male    122  108  347

In [14]:
# You code here
data.groupby(['Sex', 'Pclass']).describe()

Survived                                         Age         \
                 count  mean   std  min  25%  50%  75%  max  count   mean   
Sex    Pclass                                                               
female 1          94.0  0.97  0.18  0.0  1.0  1.0  1.0  1.0   85.0  34.61   
       2          76.0  0.92  0.27  0.0  1.0  1.0  1.0  1.0   74.0  28.72   
       3         144.0  0.50  0.50  0.0  0.0  0.5  1.0  1.0  102.0  21.75   
male   1         122.0  0.37  0.48  0.0  0.0  0.0  1.0  1.0  101.0  41.28   
       2         108.0  0.16  0.37  0.0  0.0  0.0  0.0  1.0   99.0  30.74   
       3         347.0  0.14  0.34  0.0  0.0  0.0  0.0  1.0  253.0  26.51   

               ...    Fare         Age_category                             \
               ...     75%     max        count  mean   std  min  25%  50%   
Sex    Pclass  ...                                                           
female 1       ...  134.50  512.33         85.0  1.73  0.61  1.0  1.0  2.0   
       2       ...   26.06   65.00         74.0  1.47  0.55  1.0  1.0  1.0   
       3       ...   20.22   69.55        102.0  1.26  0.47  1.0  1.0  1.0   
male   1       ...   78.46  512.33        101.0  1.95  0.67  1.0  2.0  2.0   
       2       ...   26.00   73.50         99.0  1.59  0.61  1.0  1.0  2.0   
       3       ...   10.01   69.55        253.0  1.37  0.53  1.0  1.0  1.0   

                          
                75%  max  
Sex    Pclass             
female 1       2.00  3.0  
       2       2.00  3.0  
       3       1.75  3.0  
male   1       2.00  3.0  
       2       2.00  3.0  
       3       2.00  3.0  

[6 rows x 48 columns]

* 108 men in second class were on board

**3. What are median and standard deviation of `Fare`?. Round to two decimals.**
- median is  14.45, standard deviation is 49.69
- median is 15.1, standard deviation is 12.15
- median is 13.15, standard deviation is 35.3
- median is  17.43, standard deviation is 39.1

In [15]:
# You code here
data["Fare"].agg({'median', 'std'})

std       49.69
median    14.45
Name: Fare, dtype: float64

In [16]:
data["Fare"].describe()

count    891.00
mean      32.20
std       49.69
min        0.00
25%        7.91
50%       14.45
75%       31.00
max      512.33
Name: Fare, dtype: float64

* The median is 14.45 and the standard deviation is 49.69

**4. Is that true that the mean age of survived people is higher than that of passengers who eventually died?**
- Yes
- No


In [17]:
data.groupby(['Survived']).agg({'mean'})["Age"]

mean
Survived       
0         30.63
1         28.34

* No, the mean age of the survivors is lower than the mean age of the passengers that didn't survive.

**5. Is that true that passengers younger than 30 y.o. survived more frequently than those older than 60 y.o.? What are shares of survived people among young and old people?**
- 22.7% among young and 40.6% among old
- 40.6% among young and 22.7% among old
- 35.3% among young and 27.4% among old
- 27.4% among young and  35.3% among old

In [18]:
young_or_old = lambda x: "young" if x < 30 else ("old" if x > 60 else "middle")
data["age_category"] = data["Age"].apply(young_or_old)
pd.crosstab(data.loc[data["age_category"] != "middle" ,"age_category"], data["Survived"], normalize="index")

Survived         0     1
age_category            
old           0.77  0.23
young         0.59  0.41

In [19]:
# You code here
data.loc[data["Age"] < 30, "Survived"].describe()

count    384.00
mean       0.41
std        0.49
min        0.00
25%        0.00
50%        0.00
75%        1.00
max        1.00
Name: Survived, dtype: float64

In [20]:
data.loc[data["Age"] > 60, "Survived"].describe()

count    22.00
mean      0.23
std       0.43
min       0.00
25%       0.00
50%       0.00
75%       0.00
max       1.00
Name: Survived, dtype: float64

* yes, younger people survived with ~41% chance and older people with ~23% chance 

**6. Is that true that women survived more frequently than men? What are shares of survived people among men and women?**
- 30.2% among men and 46.2% among women
- 35.7% among men and 74.2% among women
- 21.1% among men and 46.2% among women
- 18.9% among men and 74.2% among women

In [21]:
# You code here
pd.crosstab(data["Sex"], data["Survived"], normalize="index")

Survived     0     1
Sex                 
female    0.26  0.74
male      0.81  0.19

Yes, males survived only by 19%, whereas females survived with a probability of 74%

**7. What's the most popular first name among male passengers?**
- Charles
- Thomas
- William
- John

In [22]:
last = lambda x: x[-1]
data.Name.str.replace(r"^\w","").str.split(" ").apply(last).value_counts()

John          16
Henry         15
William       15
James         14
Jr             9
              ..
Hitchcock)     1
Clyde          1
Theodore       1
Thamine        1
Loraine        1
Name: Name, Length: 636, dtype: int64

In [23]:
#You code here
male_names = data.loc[data["Sex"] == "male", "Name"]
male_names = male_names.str.replace(r"^\w]", "").str.replace(r"\s.*\.", "")
male_names = male_names.str.split(" ", expand=True)
male_names = np.vstack([male_names.iloc[:,0:5]]).flatten()
male_names = male_names[male_names != '']
pd.Series(male_names).value_counts().dropna().head(5)

William    47
John       36
Henry      28
James      20
George     20
dtype: int64

* William is the most popular first name

**8. How is average age for men/women dependent on `Pclass`? Choose all correct statements:**

1.  On average, men of 1 class are older than 40
2.  On average, women of 1 class are older than 40
3.  Men of all classes are on average older than women of the same class
4.  On average, passengers ofthe first class are older than those of the 2nd class who are older than passengers of the 3rd class

In [24]:
# Average age by Sex and Pclass
data.pivot_table(["Age"], ["Pclass", "Sex"], aggfunc="mean")

Age
Pclass Sex          
1      female  34.61
       male    41.28
2      female  28.72
       male    30.74
3      female  21.75
       male    26.51

In [25]:
# Average age by Pclass
data.pivot_table(["Age"], ["Pclass"], aggfunc="mean")

Age
Pclass       
1       38.23
2       29.88
3       25.14

*   1,3,4 are correct.

## Useful resources
* The same notebook as an interactive web-based [Kaggle Kernel](https://www.kaggle.com/kashnitsky/topic-1-practice-analyzing-titanic-passengers) with a [solution](https://www.kaggle.com/kashnitsky/topic-1-practice-solution)
* Topic 1 "Exploratory Data Analysis with Pandas" as a [Kaggle Kernel](https://www.kaggle.com/kashnitsky/topic-1-exploratory-data-analysis-with-pandas)
* Main course [site](https://mlcourse.ai), [course repo](https://github.com/Yorko/mlcourse.ai), and YouTube [channel](https://www.youtube.com/watch?v=QKTuw4PNOsU&list=PLVlY_7IJCMJeRfZ68eVfEcu-UcN9BbwiX)


Author: [Yury Kashnitsky](https://yorko.github.io). This material is subject to the terms and conditions of the [Creative Commons CC BY-NC-SA 4.0](https://creativecommons.org/licenses/by-nc-sa/4.0/) license. Free use is permitted for any non-commercial purpose.